### ISSUE TITLE
Add the full version number to the installation directory name

### ISSUE THREAD

**Issue by joshdover** [elastic/elastic-agent#2579]
## Dependencies ##
- Blocked by https://github.com/elastic/elastic-agent/issues/2873
- Blocked by https://github.com/elastic/elastic-agent/issues/2582
- Blocked by ttps://github.com/elastic/elastic-agent/issues/3639

## Issue ##

We need the full version number of agent to be present in the directory we install into. This is necessary for changing how we release agent.

## Suggested solution ##

Add the full version number to the installation directory name.
Example:
```
data/
└── elastic-agent-7ac68h-8.8.1
└── elastic-agent-7ac68h-8.8.1.er1
```

## Definition of done ##

- [ ] Installation directory now includes the full version number.
- [ ] Add a new integration test [here](https://github.com/elastic/elastic-agent/blob/main/testing/integration/install_test.go) to validate that installation directory is the correct one.

**Comment by cmacknz** [comment]
Right now the upgrade process temporarily has multiple versions of the agent extracted, so it can rollback. The agent versions are installed under the /data directory with the short Git SHA as the version identifier:

data/
└── elastic-agent-383752
└── elastic-agent-7ac68h

This works fine besides being impossible to tell which version is which by manual inspection. This approach fails if we want the commit hash of the agent itself to be the same for two emergency releases. What you would want instead is something like:

data/
└── elastic-agent-7ac68h-8.8.1
└── elastic-agent-7ac68h-8.8.1.er1

This allows us to keep using the agent Git SHA in the path and also makes it obvious which version is which, which I want for convenience anyway. These `elastic-agent-xxxxx` paths now show up in our diagnostics and make debugging one step less convenient.

Another approach to this would be to make the unique identifier something other than the agent Git SHA, like computing the hash of the agent package contents at build time. This would also be unique, but the directory name wouldn't be something you could look up in Git. We would need to rely on a manifest file listing the detailed version information of each component of the agent package, but we may want that anyway.

**Comment by pchila** [comment]
Since we have to change the format of the directories and stamp out new versions, I have 1 question and (maybe) 1 proposal.
The question first: is the precedence relationship below correct? :
`8.8.0 < 8.8.1-er.1 < 8.8.1`

If the precedence is correct, can we just have the directories to follow the semver version strings like
```
data/
└── elastic-agent-8.8.1+7ac68h
└── elastic-agent-8.8.1-er.1+7ac68h

```
@cmacknz , @joshdover  what do you think ?

**Comment by jlind23** [comment]
@pchila this is correct, after discussing this today we will rather rely on changing the patch digit so having only something like:
`8.8.1 < 8.8.2 < 8.8.3 `

**Comment by jlind23** [comment]
Moving this to a later sprint as the plan changed a bit.

**Comment by cmacknz** [comment]
I would still like the version numbers in the installation path to make debugging upgrade issues easier, so there's still value in having this regardless of the plan to change how we release agent. 

Right now the agent diagnostics will contain these directory version hashes directly and you have to use Git to figure out which version is which. This is an unnecessary barrier to debugging for users who aren't developers.

**Comment by joshdover** [comment]
Agreed. @jlind23 I'm going to move this back into the current sprint, with a slightly lower priority.

**Comment by cmacknz** [comment]
We might be able to solve https://github.com/elastic/elastic-agent/issues/2780 as part of this as well to let the current version of the agent upgrade to itself. We need a way to break the path collision in this upgrade scenario which we can do by having some special convention for resolving the collision when the agent upgrades to itself.

**Comment by pchila** [comment]
Just had a quick look at this and there are a couple of elements that may change the sizing and I will report here for posterity:
- the [directory name when installing](https://github.com/elastic/elastic-agent/blob/e16fa3f1d363dd2af428726d8e35f887647f13ff/internal/pkg/agent/install/install.go#L26-L50) depends on the directory name created in the archive when packaging (since we copy over all the contents during installation).
- during packaging the directory where we place binaries inside `/data` depends on `{{.BeatName}}-{{ commit_short }}` and we have [multiple occurrences throughout `packages.yaml`](https://github.com/elastic/elastic-agent/blob/e16fa3f1d363dd2af428726d8e35f887647f13ff/dev-tools/packaging/packages.yml#L151)
- agent upgrade:
  - [unpacks everything under `data` as-is](https://github.com/elastic/elastic-agent/blob/e16fa3f1d363dd2af428726d8e35f887647f13ff/internal/pkg/agent/application/upgrade/step_unpack.go#L31-L43) retrieving the hash via `.elastic-agent.active.commit`
  -  relies heavily on the convention of `data/elastic-agent-<short hash>`  agent directories to [copy runtime data, find and switch links for upgrade/rollback](https://github.com/elastic/elastic-agent/blob/e16fa3f1d363dd2af428726d8e35f887647f13ff/internal/pkg/agent/application/upgrade/upgrade.go#L132-L164)
- throughout agent code the `VersionedHome()` is [hardcoded](https://github.com/elastic/elastic-agent/blob/e16fa3f1d363dd2af428726d8e35f887647f13ff/internal/pkg/agent/application/paths/common.go#L174-L176) to the `elastic-agent-<short commit>`
  
This means that we cannot have different directory names without some identifier of agent version that lives inside `data` directory (something very similar to what is needed for https://github.com/elastic/elastic-agent/issues/2582) and we need to launch upgrades and watchers from the new version of agent as we now need to support both directory formats as we may be upgrading all the way from back in 7.17.

To sum up:
- We probably need to develop this feature and #2582 together (not ideal, it only adds to the overall risk) or put #2582 as prerequisite of this (and not the other way around as suggested by @cmacknz )
- We need to change a very brittle build/packaging system that relies on a lot of ENV vars to communicate between build steps and abuses go templates
  - We need to introduce a separate version for agent which contains exactly the version we set during CI build/release and we need sensible defaults for local builds (what is applicable right now for beats is not good enough for agent)
  - I would suggest to have the exact same agent version (including the build ID) as part of the directory name instead of using the agent commit (we already have that in the package and we can have the full build information in a file inside `/data` directory)
- We need to support both old and new directories during upgrades 
  - install paths of both from and new directories must be included in the update marker and no longer calculated)
  - the path of the new version of the agent has to be deduced from the a marker file created during packaging or from the location of `elastic-agent` executable inside the archive
  - this will break some of the upgrade paths as old agent versions cannot cope with the new directory structure. To avoid this we can (all the options are in alternative to one another):
    - just add the support for upgrade to a different directory  **without changing the packaging directory** (8.9.0 for example) and then change the packaging in another future release (8.9.1 for example) and mark that upgrades to >8.9.0 **must** first complete an upgrade to 8.9.0 before proceeding further
    - shift the responsibility of the upgrade to the new version of the agent with the old agent spawning a subprocess after downloading and unpacking the new agent (the same problem with a mandatory specific version upgrade applies as above)
    - try to support both path formats using links, however this defeats the purpose of this issue since clashes are still possible

- We need to check all occurrences in the code of `paths.VersionedHome()` and `paths.Home()` to see if we need to duplicate logic between legacy and new directory format


This change is very risky as we don't have good test coverage for packaging, install and upgrade processes, so all of those need to be added to this and I think this makes it quite the big-bang :(

Any thoughts @joshdover , @jlind23 , @cmacknz  ?
Do we still want to put this in 8.9.0 ?
Do we introduce a breaking change for the upgrade process ? Should we phase this across multiple versions to reduce risk ?

**Comment by jlind23** [comment]
> Do we still want to put this in 8.9.0 ?
Do we introduce a breaking change for the upgrade process ? Should we phase this across multiple versions to reduce risk ?

If this product is overall not ready then we should rather lean towards postponing the change. 
We do NOT want to introduce a breaking change for the upgrade process and we shouldn't force users to upgrade to a specific version before being able to upgrade to a later one. TL;DR: 8.9 shouldn't be a pre-requisites for an upgrade to 8.9+

Happy to make https://github.com/elastic/elastic-agent/issues/2582 a pre-requisites for this and I will push this issue to later phase for now until we come up with a proper plan.

**Comment by cmacknz** [comment]
We discussed this in the agent core team meeting today. @pchila thinks he has a path forward, which depends on the implementation in https://github.com/elastic/elastic-agent/issues/2582.

* In the case that we are upgrading from a version that does not understand the new agent package version file, then we upgrade normally creating the elastic-agent-$hash directory with the short commit.
  * This works because any version that is not an ER release will always have a different hash than the ER release so there is no chance of collision. The chance of collision only happens when we allow releases to happen without changing the agent commit that is included.
  * We need to make sure we start the upgrade watcher from the new version of the agent.
* In the case that we are upgrading from a version that does understand the new agent package version file, then it can look inside the new upgrade archive and create the directory with the package version instead of the agent build hash.

This means that the agents that do understand the package version format need to handle being installed under either elastic-agent-$hash or elastic-agent-$packageVersion. We will need to wrap all of the logic that references the versioned home path to make it detect the path conditionally.

**Comment by cmacknz** [comment]
We also agreed that we will not merge any of these changes until we have each of the new upgrade cases introduced by this change covered by integration tests.

**Comment by cmacknz** [comment]
I am marking this as blocked by https://github.com/elastic/elastic-agent/issues/3639, since the solution here has become complex enough that we should step back and design it in a way that prevents us from having this or any other similar problem again in the future.

**Comment by cmacknz** [comment]
FYI the new versioning scheme we are likely to use looks like `8.11.1+build20231103` in case that affects the design at all. As before the buildXXXXX timestamp changing does not guarantee that the elastic agent commit actually changed.

**Comment by pchila** [comment]
@cmacknz During implementatino I was going for `8.11.1+git-abcdef` to somewhat maintain the git commit info.
We can easily switch to `+build<build_id>` but this implies that the buildid has to be passed from the outside to the package mage task. What buildid should we use when the build id is not passed (for example local builds) ?

**Comment by pierrehilbert** [comment]
Last time we discussed this, we thought about having this YYYYMMDD format by default and be able to override it in case we need to produce two builds the same day.

**Comment by pchila** [comment]
@pierrehilbert thanks for the clarification. This means that we need an additional BUILD_ID env var for packaging task and if empty substitute it with the current YYYYMMDD
Just one note: currently the elastic-agent build ids are UIDs like `8.11.0-d2c52cf9`. Using an override we can get something similar if we pass the buildid in from the build pipeline and maybe we want to do that until the release process of elastic-agent changes...

**Comment by cmacknz** [comment]
>  This means that we need an additional BUILD_ID env var for packaging task and if empty substitute it with the current YYYYMMDD


Just to clarify, I wanted to double check that the agent, package version file, and new upgrade path can work with the proposed YYYYMMDD format. You don't actually need to change the build system to do this as part of this work, that is going to be tracked and changed separately.

**Comment by pchila** [comment]
@cmacknz the agent package version can be whatever string we like, however the work in progress we already have for this feature creates the paths like `elastic-agent-<package version>+git<elastic-agent repo commit sha>` (the git commit is part of the build metadata information if we interpret the string as semver)

Furthermore the upgrade to a specific snapshot build for agent supports strings like `8.11.0-SNAPSHOT+<build id>` which is also different from the proposed format.

One **hacky** way to get something similar to the format `8.11.1+build20231103` for the directory names and the version is to pass the string `x.y.z+buildYYYYMMDD` as AGENT_PACKAGE_VERSION like:
```
 AGENT_PACKAGE_VERSION="8.12.0+build20231103" EXTERNAL=true mage package
 ```
however that interferes with the `SNAPSHOT` flag, the version will look strange, so for

```
 AGENT_PACKAGE_VERSION="8.12.0+build20231103" EXTERNAL=true mage package
```
we get
```
➜  distributions git:(add-version-to-agent-installation-directory) ✗ tree -L 3
.
├── elastic-agent-8.12.0+build20231102-SNAPSHOT-linux-x86_64
│   ├── data
│   │   ├── elastic-agent-8.12.0+build20231102+git-c3c63c
│   │   └── elastic-agent-c3c63c
│   ├── elastic-agent -> data/elastic-agent-8.12.0+build20231102+git-c3c63c/elastic-agent
│   ├── elastic-agent.reference.yml
│   ├── elastic-agent.yml
│   ├── LICENSE.txt
│   ├── NOTICE.txt
│   └── README.md
├── elastic-agent-8.12.0+build20231102-SNAPSHOT-linux-x86_64.tar.gz
└── elastic-agent-8.12.0+build20231102-SNAPSHOT-linux-x86_64.tar.gz.sha512

4 directories, 8 files

➜  distributions git:(add-version-to-agent-installation-directory) ✗ ./elastic-agent-8.12.0+build20231102-SNAPSHOT-linux-x86_64/elastic-agent version --binary-only
Binary: 8.12.0+build20231102-SNAPSHOT (build: c3c63cd9bf828b32fd2a276ed7838594805028d5 at 2023-11-02 15:15:11 +0000 UTC)
➜  distributions git:(add-version-to-agent-installation-directory) ✗ ./elastic-agent-8.12.0+build20231102-SNAPSHOT-linux-x86_64/elastic-agent version --binary-only --yaml
binary:
  version: 8.12.0+build20231102
  commit: c3c63cd9bf828b32fd2a276ed7838594805028d5
  build_time: 2023-11-02T15:15:11Z
  snapshot: true
```
(Disregard the `+git...` part of the directory name as that has been added as part of this development and can be easily removed)

I guess this is not what we want, at least for snapshot builds...

**Comment by cmacknz** [comment]
I don't think we'll see `8.12.0+build20231103-SNAPSHOT` versions in practice, those will be from builds produced on a weekly schedule similar to snapshots but they are intended to only contain releasable changes.

**Comment by cmacknz** [comment]
I have been thinking more about this recently. In particular about how we could solve https://github.com/elastic/elastic-agent/issues/2579 without introducing significant additional complexity to the upgrade process, or at least introducing complexity that has to interact with previous versions of the agent where unexpected interactions cannot easily be fixed.

We are trying to support this for a release strategy where the elastic-agent binary does not change between upgrades and the 6 character hash in `data/elastic-agent-$hash` collides when this future upgrade is attempted. Here is a sequence of examples that illustrate the problem:

1. Install v7.17.15 containing the following files:
```
    .elastic-agent.active.commit
        abcd
    data/elastic-agent-abcd
        elastic-agent v7.17.15
        /components
            endpoint-security v7.17.15
```
2. Upgrade to v8.15.0 containing the following files:
```
    .elastic-agent.active.commit 
        efgh
    data/elastic-agent-efgh
        elastic-agent v8.15.0
        /components
            endpoint-security v8.15.0
```
3. Upgrade to v8.15.0+build20231130 containing the following files:
```
    .elastic-agent.active.commit
        efgh
    data/elastic-agent-efgh
        elastic-agent v8.15.0
        /components
            endpoint-security+git789 v8.15.0+build20231130

```

In this example the `v8.15.0+build20231130` release only changes the endpoint-security version and the `efgh` hash in `data/elastic-agent-efgh` collides between the original `v8.15.0` version and the `v8.15.0+build20231130` version.

A key observation is that the elastic-agent binary version is guaranteed to be different until we upgrade between two versions that were produced after this new upgrade strategy is implemented. So in every currently released agent version and all future versions up to the first release where the elastic-agent binary version does not change, a hash collision during upgrades is impossible. In this specific example `v7.17.15` can successfully upgrade to either the future `v8.15.0` or `v8.15.0+build20231130` version if we change nothing. The problem only exists in the upgrade from the `v8.15.0` and `v8.15.0+build20231130` version which do not exist yet.

All current discussions around how to solve this problem involve the future agent version changing the directory structure and doing something to help previous versions migrate to it. With the observation above that all previous versions can successfully upgrade to any future version if we change nothing, I think we should look at this differently. **Instead of helping previous versions migrate to a new structure, we should leave the structure alone and explore having the future versions have a way to resolve this conflict with other future versions.**

A proposed approach that I think is simpler to implement follows. The changes can all be made on `main`, do not affect previous versions at all, and the observations above mean we are guaranteed to be able to upgrade to a version of the agent that contains these changes successfully from any previously released version.

1. Continue to package the agent using the `elastic-agent-$hash` pattern with `$hash` being the first 6 characters read from the `.elastic-agent.active.commit` file. This guarantees all existing agent versions can upgrade with no changes that interact with previously released versions.
2. We allow the agent to work with arbitrary `data/elastic-agent-$pattern` directory structures without an arbitrary character limit. @pchila has a draft RFC that proposes adding a metadata file that future versions read but previous versions will ignore. This file could contain the destination directory we should extract the agent to using a naming scheme that does not collide and has no length limit. 
3. One approach to resolve collisions is then to extract the new agent package into a sub-directory that cannot have a collision (for example `data/elastic-agent-candidate`), read the metadata file above to determine the destination directory, and then move the agent to its final destination. This will work but involves extracting and then moving the agent. It also lets the upgrade metadata instruct the agent to extract to arbitrary location which may be a negative feature for security reasons.
4. Another approach is to try to rewrite the `data/elastic-agent-$hash` directory name to `elastic-agent-$hash-$version` as we are extracting it. `$version` is the version from the upgrade action or package name, this assumes all `$hash-$version` combinations are unique. To support the possibility of downgrades we could have the agent try to exec the new agent version from both `data/elastic-agent-$hash` and `data/elastic-agent-$hash-$version` and only fail if neither exists. Both the zip and tar packages allow us to inspect files names before we write them to disk and in theory we could rewrite them:

https://github.com/elastic/elastic-agent/blob/f7dcbd74ce0e27d3b46b478755168c66d78dc111/internal/pkg/agent/application/upgrade/step_unpack.go#L67-L68

https://github.com/elastic/elastic-agent/blob/f7dcbd74ce0e27d3b46b478755168c66d78dc111/internal/pkg/agent/application/upgrade/step_unpack.go#L163-L164

**Comment by pchila** [comment]
@cmacknz here are my thoughts on your proposal:

> Continue to package the agent using the elastic-agent-$hash pattern with $hash being the first 6 characters read from the .elastic-agent.active.commit file. This guarantees all existing agent versions can upgrade with no changes that interact with previously released versions.

This is true for all versions of elastic-agent released unitl now because the unified release process always introduce a new commit. If I understand this correctly you want to keep *releasing* agent with the current package structure, correct?

> On main we allow the agent to work with arbitrary data/elastic-agent-$pattern directory structures without an arbitrary character limit. @pchila has a draft RFC that proposes adding a metadata file that future versions read but previous versions will ignore. This file could contain the destination directory we should extract the agent to using a naming scheme that does not collide and has no length limit.

When you say on `main` are you referring just to SNAPSHOT versions or maybe all future versions once this change has been introduced ? 
Would there be a difference between a SNAPSHOT and a release package structure?
What would be the expected directory structure for a clean install and for an upgrade ?

It's not clear to me how this is supposed to work.

> Following from this one approach to resolve collisions is then to extract the new agent package into a sub-directory that cannot have a collision (for example data/elastic-agent-candidate), read the metadata file above to determine the destination directory, and then move the agent to its final destination. This will work but involves extracting and then moving the agent. It also lets the upgrade metadata instruct the agent to extract to arbitrary locations which is positive for future proofing but may be a negative feature for security.

This implies that the upgrade (and probably also the install for consistency) process will change starting with the PR(s) for this issue and the backward compatibility is maintained only though the package structure?
As for security about the metadata in moving the agent, we can force relative paths that fall inside the agent top directory (this could still rewrite existing agents installations in case of collisions)

Would this happen only if the agent in control of the upgrade is a new one (at least a version that is aware of the manifest file) ?

> Another approach is to try to rewrite the data/elastic-agent-$hash directory name to elastic-agent-$hash-$version as we are extracting it. $version is the version from the upgrade action or package name, this assumes all $hash-$version combinations are unique. To support the possibility of downgrades we could have the agent try to exec the new agent version from both data/elastic-agent-$hash and data/elastic-agent-$hash-$version and only fail if neither exists. This avoids moving the agent but we don't have arbitrary control over the destination directory. Both the zip and tar packages allow us to inspect files names before we write them to disk and in theory we could rewrite them:
...
Once all existing agent versions that need to work with elastic-agent-$hash reach end of support some unknown point in the future, we could start packaging the agent using data/elastic-agent-$hash-$version by default. This approach also means the data/elastic-agent-* pattern used is not always consistent and depends on the version we upgraded from.

Who has the responsibility for renaming? The upgrade process or the new version of agent itself ? As for downgrades, depending on which version is in control of extracting the new files we may have different results, for example if the installed agent is in control and 8.12 is an agent version where this change has been implemented:
```
Initial agent 7.17 -> /opt/Elastic/Agent/elastic-agent-<hash717>
upgrade to 8.12.0 without rename -> /opt/Elastic/Agent/elastic-agent-<hash812>
downgrade back to 7.17 -> /opt/Elastic/Agent/elastic-agent-<hash717>-7.17.X  <- probably broken
```
Are we gonna gate the feature so that we maintain the old behavior for old versions in case of downgrade ? If that's the case it means that the agent has to "know" which versions can be renamed with the uncertainty only for SNAPSHOT versions of the <major.minor.patch>  where the change has been introduced :thinking: 
The main difference of this approach compared to the other one is that it does not need a manifest file, correct ? I feel like it's better get the version/destination path from a file contained in the package rather than extrapolating it from the upgrade command or the archive file name

**Comment by cmacknz** [comment]
> This is true for all versions of elastic-agent released unitl now because the unified release process always introduce a new commit. If I understand this correctly you want to keep releasing agent with the current package structure, correct?

Yes, our build system should continue to default to the current package structure of `data/elastic-agent-$hash`.

> When you say on main are you referring just to SNAPSHOT versions or maybe all future versions once this change has been introduced ?
Would there be a difference between a SNAPSHOT and a release package structure?
What would be the expected directory structure for a clean install and for an upgrade ?

All future versions. There is no difference between a SNAPSHOT and a release in terms of package structure. A SNAPSHOT is just what we call our daily builds. The difference is in the expected stability and support contract.

> This implies that the upgrade (and probably also the install for consistency) process will change starting with the PR(s) for this issue and the backward compatibility is maintained only though the package structure?

Yes, backwards compatibility is maintained by keeping the old package structure.

> Would this happen only if the agent in control of the upgrade is a new one (at least a version that is aware of the manifest file) ?

Yes that is a key part of this. If we do not change the package structure, every old agent is guaranteed to be able to upgrade to a new one without issue, and new agents are aware of this version conflict and manifest file.

> Who has the responsibility for renaming? The upgrade process or the new version of agent itself ?

The rename has to happen during the unpack step and involve moving or renaming the directory containing the new elastic-agent executable, which likely means this logic has to stay in the version we are upgrading from. So the download+unpack+rename step has to stay in the "current" agent and the rest of the upgrade could be done by the "next" agent.

> Initial agent 7.17 -> /opt/Elastic/Agent/elastic-agent-<hash717>
upgrade to 8.12.0 without rename -> /opt/Elastic/Agent/elastic-agent-<hash812>
downgrade back to 7.17 -> /opt/Elastic/Agent/elastic-agent-<hash717>-7.17.X  <- probably broken

We can have logic to only do the rename if there is a conflict detected, and we are guaranteed that no old version that doesn't know about the new path structure will ever have a conflict (or we could have the logic conditional on the version we are upgrading to since we have that information). SO if we downgraded from 8.12.0 to 7.17 it should work because the directory structure for the 7.17 agent is unchanged.

> Are we gonna gate the feature so that we maintain the old behavior for old versions in case of downgrade ? If that's the case it means that the agent has to "know" which versions can be renamed with the uncertainty only for SNAPSHOT versions of the <major.minor.patch> where the change has been introduced 🤔

Yes we will need something to handle downgrades. See above. It is likely simpler to only activate the rename logic if we detect there is a version conflict during the unpack step, because we are guaranteed that all older versions that can't handle conflicts during upgrade will themselves not have a conflict.

> The main difference of this approach compared to the other one is that it does not need a manifest file, correct ? I feel like it's better get the version/destination path from a file contained in the package rather than extrapolating it from the upgrade command or the archive file name

Step 4 as I wrote it doesn't require a manifest file and doesn't require us to unpack the agent and then move/copy it. Although now that I am thinking about it we could do a blend of 3 and 4 where we keep the manifest file, unpack to a known directory that doesn't collide like `elastic-agent-next` by rewriting the `elastic-agent-$hash` name on the file, then read the file, and rename the directory. That would avoid the performance cost of having to move the agent.

**Comment by blakerouse** [comment]
I think that @cmacknz is correct on this logic and I think this is the best path forward as well.

I would like to explore the removal (not adding) of the manifest file. We manage the entire extraction process in the Elastic Agent (its not shelled out to tar or anything). We can easily see early from parsing the zip or tar file that a collision will happen, at that point  the Elastic Agent can do what is needed to ensure that it doesn't overwrite its current execution directory. `elastic-agent version` can be called from the binary that is extracted to get the version information without requiring another file to parse.

**Comment by cmacknz** [comment]
It is technically possible to do this without a manifest file. However I would like us to avoid hard coding the directory convention we use to avoid conflicts into the agent we are upgrading from, because if it turns out we made a mistake we have no way to fix it.

Having this name convention reside in a manifest file that is included in new agent we are downloading is a way to solve this problem. This is inline with https://github.com/elastic/elastic-agent/issues/3639 where we should move as much as possible to the new agent we are upgrading to.

Having a manifest file also gives us an extension point for parameterizing other parts of the upgrade process as needed, for example we can include the full path to the new agent binary, the arguments to pass to it, etc. I think this helps future proof the upgrade process and gives us some nice extension points for https://github.com/elastic/elastic-agent/issues/3639. 

For example if we were to rely on parsing the version out of `elastic-agent version` but we accidentally broke that command in some way, having the command used to obtain the version in a manifest file would give us a way out of this as we could configure the old agent do something else for example.

**Comment by blakerouse** [comment]
> It is technically possible to do this without a manifest file. However I would like us to avoid hard coding the directory convention we use to avoid conflicts into the agent we are upgrading from, because if it turns out we made a mistake we have no way to fix it.

Don't see how it would result in hard-coding the directory convention. We could be smart in finding the binary, we need to find the binary anyway. I can see how a manifest would make it more defined and remove the need for searching in the case something changed.

> 
> Having this name convention reside in a manifest file that is included in new agent we are downloading is a way to solve this problem. This is inline with #3639 where we should move as much as possible to the new agent we are upgrading to.
> 
> Having a manifest file also gives us an extension point for parameterizing other parts of the upgrade process as needed, for example we can include the full path to the new agent binary, the arguments to pass to it, etc. I think this helps future proof the upgrade process and gives us some nice extension points for #3639.

Don't know why you would pass parameters to something that would be hard coded into say the `upgrade.exe`. If the parameters are required then it should just be something that `upgrade.exe` does.

> 
> For example if we were to rely on parsing the version out of `elastic-agent version` but we accidentally broke that command in some way, having the command used to obtain the version in a manifest file would give us a way out of this as we could configure the old agent do something else for example.

That logic applies the same if you break the manifest file or the content of the manifest file is incorrect. Don't see as negating any more of a chance of the manifest breakage, and something that should also be covered in testing. The manifest file will also have to be covered in testing, resulting in basically the same reasoning.

**Comment by blakerouse** [comment]
I am just trying to find a path where its just extremely simple. A super easy path would be:

1. Extract into the same directory always like `data/upgrade-candidate`.
2. Call `data/upgrade-candidate/upgrade.exe`.

`upgrade.exe` does all the work from there and knows what to do. In the case of an old Elastic Agent you make the directory structure work as stated in above which I like (where it will ignore the manifest anyway).
